ColorPicker UI improvements after adding intensity slider.

WhalesState opened the pull request:
**changes:**

- Sorted intensity slider to appear after the alpha slider for consistency with other slider styles.

- Added icon for hex code input, matching the new script code icon style.

- Modes are now [RGB, HSV, Linear, OKHSL], displayed horizontally without an extra popup for OKHSL.

- Moved colorize sliders option to theme constants (purely visual change) and removed the mode `PopupMenu`.

- Added editor setting: `color_picker_colorize_sliders`.

- Changed the `label_width` from 10 to 16 to prevent the sliders from moving when changing modes.

- Changed the sampler to use `Control` node instead of `TextureRect` since it 's not using the `texture` property.

- Removed the `Hex/Expr` Label, maybe we can provide a better tooltip text for the hex button instead.

beicause commented:
Based on the PR description, most of the changes look good. Some personal feedback from my side:

>Reverted Linear mode renaming back to RAW mode for consistent short names.
>Modes are now [RGB, HSV, RAW, HSL], displayed horizontally without an extra popup for OKHSL.

I disagree with renaming "Linear" and "OKHSL" to "RAW" and "HSL". They are not the same - what is "RAW"? and "OKHSL" is different from traditional "HSL". It's not good to sacrifice accuracy for the sake of shorter names, which may confuse users.

>Sorted intensity slider to appear after the alpha slider for consistency with other slider styles.

In my opinion, placing the intensity slider before the alpha slider is easier to use. It's inconvenient if the unchanged alpha slider is in the middle.

>Fixed regressions from previous intensity slider changes.

What other specific regressions are also fixed?

WhalesState commented:
* Moved the bug fixes to another PR: https://github.com/godotengine/godot/pull/110160

beicause commented:
>After the intensity PR, disabling the intensity sliders reverts the behavior back to what was previously called RAW, where values greater than 1.0 can still be set. For that reason, calling it Linear wouldn’t really be accurate either. I decided to keep the existing RAW name, since it has been documented and consistently used for years without causing confusion.

It's not about whether the intensity slider is enabled or not, it's about color space. If you set the linear color sliders to 0.5, the actual color is 0.735, not it's not raw.
- See https://github.com/godotengine/godot/pull/103583#issuecomment-2703990932

>Aseprite itself uses the short label HSL for the Hue/Saturation/Lightness sliders,

If you want to copy the values of other HSL color to OKHSL, you will find the colors are not the same, so it's better to specify the color model clearly.

WhalesState commented:
I didn't even realize that Linear mode is not the same as RGB like before, Reverted the change, only disabled the translation for it since other modes doesn't have translatable names.

Other changes have related to intensity slider are moved to different PRs:
* https://github.com/godotengine/godot/pull/110386
* https://github.com/godotengine/godot/pull/110384

WhalesState commented:
> That's because they aren't even words. The Linear was marked for translation in #109354.

If we allow translating all of them then some other languages may find a workaround for all mode names but for Arabic it's impossible to translate `RGB`, `HSV` or `OKHSL`.

<img width="321" height="137" alt="image" src="https://github.com/user-attachments/assets/e5e8bee8-b9e9-44fd-9e3b-a383e4d5fcc9" />

I will revert this change for now since I only use the editor in English, Just pointing out that it will look weird for all RTL languages.

Calinou reviewed (commented):
Tested locally, it works as expected. Code looks good to me.

In terms of usability, I think this is a pretty nice improvement already. I prefer the new location of the intensity slider too, it's easier on muscle memory when switching between no-alpha and with-alpha color pickers.

Some feedback:

- When switching the editor to a light theme, the "color picker type" icon in the right isn't updated:

<img width="536" height="135" alt="image" src="https://github.com/user-attachments/assets/7ab433b4-3dcc-43be-8dd0-047351014839" />

All color picker icon types in the dropdown are also affected:

<img width="302" height="317" alt="image" src="https://github.com/user-attachments/assets/b317b330-6ed1-4abc-aa32-dd0bc3191f30" />

In both cases, this happens even after restarting the editor, so the issue isn't in a lack of `NOTIFICATION_THEME_CHANGED` usage. It's either the icon file having an incorrect color, or the icon being incorrectly listed in the icon conversion exclusions.

- Pressing <kbd>Ctrl + Enter</kbd> on the color text field should focus it and select all its text, so you can enter another color code or name afterwards without needing to click with the mouse.

- Some icons don't scale up for hiDPI displays in the editor, like the context menu icons next to the Swatches section:

<img width="196" height="263" alt="image" src="https://github.com/user-attachments/assets/562f2a85-c095-4459-baa4-c2b2428f0f8b" />

As well as the "Swatches" and "Recent colors" chevrons:

<img width="85" height="117" alt="image" src="https://github.com/user-attachments/assets/30c75a99-b93e-4fff-93f4-5bf8fe093c88" />

- The Swatches context menu icon should also have a square minimum size, so that its clickable area is always square:

<img width="63" height="66" alt="image" src="https://github.com/user-attachments/assets/f1688111-b89b-42f7-875b-920f8379bbd9" />

___

For future PRs:

- We could make the value slider slightly rounded to match the focus StyleBox, so that it doesn't poke out of the focus StyleBox like it does currently:

<img width="52" height="28" alt="image" src="https://github.com/user-attachments/assets/3b21bad5-4fb8-485a-9d77-e110d9bac8bb" />

This also applies to the corners of rectangle color pickers:

<img width="39" height="15" alt="image" src="https://github.com/user-attachments/assets/b2e46f6d-61cf-4f9f-8372-6852a011593c" />

Alternatively, we can use special focus StyleBoxes that don't have rounded corners for those (it should be easier to implement).

- The focus outline for the color picker circle could be antialiased:

<img width="96" height="56" alt="image" src="https://github.com/user-attachments/assets/68e44464-19eb-491c-9bd6-92378aa6a213" />

The main circle is already antialiased, although it should perhaps use a slightly wider feather to improve quality.

WhalesState commented:
> * When switching the editor to a light theme, the "color picker type" icon in the right isn't updated:
> All color picker icon types in the dropdown are also affected:
> In both cases, this happens even after restarting the editor, so the issue isn't in a lack of `NOTIFICATION_THEME_CHANGED` usage. It's either the icon file having an incorrect color, or the icon being incorrectly listed in the icon conversion exclusions.

Fixed, for shape menu icons, the files were using the wrong color, for the Palette menu button, i can't find why it's not using the editor theme colors, i already have added the missing overrides in EditorThemeManager and it didn't update the icons.

> * Pressing Ctrl + Enter on the color text field should focus it and select all its text, so you can enter another color code or name afterwards without needing to click with the mouse.

It works as expected without making any changes on fedora wayland.

> * Some icons don't scale up for hiDPI displays in the editor, like the context menu icons next to the Swatches section:
> As well as the "Swatches" and "Recent colors" chevrons:

I don't know if they do scale or not, hard to see the difference since they use 16x16 size by default which will look pixelated even when scaled.

> * The Swatches context menu icon should also have a square minimum size, so that its clickable area is always square:

Fixed.

> For future PRs:
> * We could make the value slider slightly rounded to match the focus StyleBox

Made the `picker_focus_rectangle` style have sharp corners for editor, and i think we can reuse it to override other Controls focus style that have issues with the default focus style.

> * The focus outline for the color picker circle could be antialiased

Enabled antialiasing for the circle shapes.

> The main circle is already antialiased, although it should perhaps use a slightly wider feather to improve quality.

Using the default 1.0 anti-aliasing size to avoid those artifacts.

<img width="1384" height="764" alt="image" src="https://github.com/user-attachments/assets/7b2da471-b728-42e4-b802-784c0d44f263" />

Another issue: Since the corner radius of the circle shape focus is set to 256px and we do allow changing the shape width and height, the shape appears deformed when the circle shapes are not perfect and also input doesn't work correctly.

<img width="886" height="595" alt="image" src="https://github.com/user-attachments/assets/039f2775-68e2-42d4-acb7-a33ddf866bec" />

Calinou reviewed (commented):
Works great now 🙂 

I noticed the "revert to old color" icon gets converted when using a light theme in the editor, but it shouldn't, as it's always drawn over a background that does not depend on the theme color:

<img width="35" height="28" alt="Screenshot 2025-09-11 at 14 55 59" src="https://github.com/user-attachments/assets/b533e4c8-75ce-4c97-8622-3ad3efa779b7" />

Its modulation is set according to the background color - if it's light enough, the icon gets darkened, otherwise it's drawn with the standard color (which is a light icon for dark theme).

This issue is also present in `master` so it's not critical to fix here, but it may be worth a look at nonetheless.

KoBeWi commented:
I'm not sure about removing the mode button completely. If we add more color modes, it will have to be re-added.
Then again, it's not needed right now, so removing it makes sense.

KoBeWi commented on scene/gui/color_picker.h:
```diff
@@ -317,6 +311,7 @@ class ColorPicker : public VBoxContainer {
 		int sv_height = 0;
 		int sv_width = 0;
 		int h_width = 0;
+		int colorize_sliders = 0;
```
```suggestion
		bool colorize_sliders = false;
```
The underlying constant can be bool, even if it still shows as integer.

KoBeWi replied:
You forgot `false`.

KoBeWi commented on scene/gui/color_picker.cpp:
```diff
@@ -139,34 +138,66 @@ void ColorPicker::_notification(int p_what) {
 
-			_reset_sliders_theme();
+			Size2 cursor_size = (theme_cache.picker_cursor->get_size() * 0.5).round();
+			real_vbox->add_theme_constant_override(SNAME("separation"), cursor_size.height);
```
This doesn't work well.

<img width="181" height="414" alt="Image" src="https://github.com/user-attachments/assets/6d2c7057-7d4b-4509-becc-22819d31440c" />

It makes sense for the shape margins, but not for the whole ColorPicker. You could add a MarginContainer for this effect (if there isn't one that can be used already).

WhalesState committed:
ColorPicker: UI improvements and bug fixes.

WhalesState commented:
Note: I still have more local changes and fixes for the `ColorPicker` that are hard to applied all together, so I'm waiting until the current PRs we have that changes `color_picker.cpp` to be merged first.

Repiteo commented:
Needs another rebase

allenwp commented:
> The intensity slider uses a different style from the colored sliders. Putting it in the middle (before alpha) breaks the visual flow and makes the layout feel inconsistent. Since intensity is also a less frequently used control, I felt it was better placed after alpha, so the main color sliders stay grouped together.

I understand what you're describing here: the layout is more visually pleasing by grouping the alpha slider with the other colour sliders.

I also disagree that the change should be made to the position of the intensity slider because this negatively affects usability: The intensity slider directly affects the colour defined by RGB, HSV, and HSL and has nothing to do with the alpha value. Placing the alpha slider after all of the colour-related sliders is an intentional GUI design choice that makes the alpha transparency distinct from the colour. Conversely, the intensity slider is placed beside the RGB, HSV, and HSL sliders because it directly changes the RGB, HSV, and HSL values. The original ordering was purposefully designed to favour usability and clarity of function, which I believe is more valuable than visual flow in this case.

The new Hex icon is a nice improvement. I'm on the fence about always showing OKHSL because it's already a visually complex tool for new users...
